### ISSUE TITLE
Very High RAM Usage

### ISSUE THREAD

**Issue by aditya23043** [YaLTeR/niri#321]
about 80-90% RAM usage according to btm (bottom system monitor) on niri alone

### System Information

* niri version: 0.1.5

* Distro: Archlinux (6.8.7-arch1-2)

* GPU: Nvidia RTX3060

* CPU: Ryzen 9 5900HX

**Comment by YaLTeR** [comment]
Huh, how much RAM do you have in total? For me niri is using around 33 MB of RAM (reserved minus shared).

**Comment by YaLTeR** [comment]
What does `htop` say for niri in columns VIRT, RES, SHR?

**Comment by aditya23043** [comment]
niri now at 19%

$ btm
![image](https://github.com/YaLTeR/niri/assets/66688960/87f0f92d-4a23-4e47-88dc-0151572be804)

$ htop
![image](https://github.com/YaLTeR/niri/assets/66688960/34f169a8-30d5-4e53-8fcd-5ba8d425f895)

**Comment by aditya23043** [comment]
31% RAM usage

![image](https://github.com/YaLTeR/niri/assets/66688960/bddbeeed-a60d-474d-981a-f311eec89f16)

**Comment by YaLTeR** [comment]
What's going on with child processes? Are you by chance spamming invalid spawn requests and triggering #229? Anything in the niri log? There shouldn't be any children left over.

**Comment by YaLTeR** [comment]
niri prints logs to stdout/stderr. If using systemd, you can find them in `journalctl -b /usr/bin/niri`

**Comment by aditya23043** [comment]
```
  ❯ journalctl -b | grep -i niri
Apr 30 14:54:50 archlinux sudo[170900]:      adi : TTY=tty2 ; PWD=/home/adi ; USER=root ; COMMAND=/usr/bin/pacman --sync -- niri
Apr 30 14:54:50 archlinux pacman[170905]: Running 'pacman --sync -- niri'
Apr 30 14:54:51 archlinux pacman[170905]: reinstalled niri (0.1.5-3)
Apr 30 15:37:32 archlinux sudo[176789]:      adi : TTY=pts/0 ; PWD=/home/adi ; USER=root ; COMMAND=/usr/bin/find / -iname niri
Apr 30 15:51:31 archlinux sudo[178277]:      adi : TTY=pts/0 ; PWD=/home/adi ; USER=root ; COMMAND=/usr/bin/find / -iname niri
Apr 30 15:56:01 archlinux sudo[178549]:      adi : TTY=pts/3 ; PWD=/home/adi ; USER=root ; COMMAND=/usr/bin/find / -iname niri

  ❯ journalctl -b /usr/bin/niri
-- No entries --
```

**Comment by YaLTeR** [comment]
Hm, I see that the parent of niri is `fish`, are you running from TTY directly? Then just redirect the output into a file with `niri --session >niri.log 2>&1` and check it afterwards.

**Comment by aditya23043** [comment]
yes, i am running from the TTY
Alright, I'll check
And also, thanks for the prompt responses

**Comment by aditya23043** [comment]
```
$ cat niri.log
[2m2024-04-30T10:36:03.973712Z[0m [32m INFO[0m [2mniri[0m[2m:[0m starting version 0.1.5 (unknown commit)
[2m2024-04-30T10:36:04.013094Z[0m [34mDEBUG[0m [2mniri_config[0m[2m:[0m loaded config from "/home/adi/.config/niri/config.kdl"
[2m2024-04-30T10:36:04.146127Z[0m [32m INFO[0m [2mniri::backend::tty[0m[2m:[0m using as the render node: "/dev/dri/renderD128"
[2m2024-04-30T10:36:04.445544Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m device added: 57857 "/dev/dri/card1"
[2m2024-04-30T10:36:04.553388Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m this is the primary node
[2m2024-04-30T10:36:04.569646Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m device changed: 57857
[2m2024-04-30T10:36:04.595517Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m connecting connector: eDP-1
[2m2024-04-30T10:36:04.595584Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m picking mode: Mode { name: "1920x1080", clock: 354730, size: (1920, 1080), hsync: (1968, 2000, 2180), vsync: (1083, 1088, 1130), hskew: 0, vscan: 0, vrefresh: 144, mode_type: ModeTypeFlags(PREFERRED | DRIVER) }
[2m2024-04-30T10:36:04.595654Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m set max bpc to 8
[2m2024-04-30T10:36:04.599830Z[0m [34mDEBUG[0m [2mniri::niri[0m[2m:[0m putting output eDP-1 at x=0 y=0
[2m2024-04-30T10:36:04.599936Z[0m [34mDEBUG[0m [2mniri::backend::tty[0m[2m:[0m device added: 57856 "/dev/dri/card0"
[2m2024-04-30T10:36:04.600864Z[0m [33m WARN[0m [2mniri::backend::tty[0m[2m:[0m error adding device: DRM access error: Error loading resource handles on device `Some("/dev/dri/card0")` (Operation not supported (os error 95))

Caused by:
    Operation not supported (os error 95)
[2m2024-04-30T10:36:04.600921Z[0m [32m INFO[0m [2mniri[0m[2m:[0m listening on Wayland socket: wayland-1
[2m2024-04-30T10:36:04.600933Z[0m [32m INFO[0m [2mniri[0m[2m:[0m IPC listening on: /run/user/1000/niri.wayland-1.179428.sock
/bin/sh: line 1: dinitctl: command not found
 
```

**Comment by YaLTeR** [comment]
Hm, so you built without systemd feature and with dinit feature?

**Comment by aditya23043** [comment]
I am not sure about that
I am not well versed with systemd or any init program

**Comment by aditya23043** [comment]
If you are talking about building niri, I just downloaded it from the AUR

**Comment by YaLTeR** [comment]
Well that's weird because neither niri-git nor niri-bin from the AUR set the dinit feature as far as I understand. So niri should not call dinitctl.

**Comment by aditya23043** [comment]
I got the normal "niri" package

```
  ❯ paru -Ss niri
extra/niri 0.1.5-3 [0 B 19.86 MiB] [Installed]
    A scrollable-tiling Wayland compositor
aur/niri-bin 0.1.5-2 [+5 ~1.37]
    A scrollable-tiling Wayland compositor
aur/niri-git 0.1.5.r15.g8d99e3c-2 [+5 ~0.98]
    Scrollable-tiling Wayland compositor

  ~ 
  ❯ paru -Si niri
Repository      : extra
Name            : niri
Version         : 0.1.5-3
Description     : A scrollable-tiling Wayland compositor
Architecture    : x86_64
URL             : https://github.com/YaLTeR/niri
Licenses        : GPL-3.0-or-later
Groups          : None
Provides        : None
Depends On      : cairo  gcc-libs  glib2  glibc  libinput  libpipewire
                  libxkbcommon  mesa  pango  pixman  seatd  systemd-libs
                  xdg-desktop-portal-impl
Optional Deps   : alacritty: a suggested GPU-accelerated terminal emulator
                  bash: for niri-session script
                  fuzzel: a suggested Wayland application launcher
                  mako: a suggested Wayland notification daemon
                  org.freedesktop.secrets: for apps to rely on secrets portal
                  swaybg: a suggested Wayland wallpaper tool
                  waybar: a suggested Wayland customizable desktop bar
                  waylock: a suggested Wayland screen locker
                  xdg-desktop-portal-gtk: a suggested XDG desktop portal
                  xdg-desktop-portal-gnome: a XDG desktop portal required for
                  screencasting
Conflicts With  : None
Replaces        : None
Download Size   : 5.04 MiB
Installed Size  : 19.86 MiB
Packager        : David Runge <dvzrv@archlinux.org>
Build Date      : Sun Apr 28 21:11:55 2024
Validated By    : SHA-256 Sum  Signature
```

**Comment by aditya23043** [comment]
should I install one of the other two?

**Comment by aditya23043** [comment]
After installing niri-bin
```
  ❯ cat niri.log 
2024-04-30T10:50:16.973529Z  INFO niri: starting version 0.1.5 (6a80078)
2024-04-30T10:50:17.017554Z DEBUG niri_config: loaded config from "/home/adi/.config/niri/config.kdl"
2024-04-30T10:50:17.184577Z  INFO niri::backend::tty: using as the render node: "/dev/dri/renderD128"
2024-04-30T10:50:17.590825Z DEBUG niri::backend::tty: device added: 57856 "/dev/dri/card0"
2024-04-30T10:50:17.592175Z  WARN niri::backend::tty: error adding device: DRM access error: Error loading resource handles on device `Some("/dev/dri/card0")` (Operation not supported (os error 95))

Caused by:
    Operation not supported (os error 95)
2024-04-30T10:50:17.592223Z DEBUG niri::backend::tty: device added: 57857 "/dev/dri/card1"
2024-04-30T10:50:17.727498Z DEBUG niri::backend::tty: this is the primary node
2024-04-30T10:50:17.748261Z DEBUG niri::backend::tty: device changed: 57857
2024-04-30T10:50:17.772279Z DEBUG niri::backend::tty: connecting connector: eDP-1
2024-04-30T10:50:17.772347Z DEBUG niri::backend::tty: picking mode: Mode { name: "1920x1080", clock: 354730, size: (1920, 1080), hsync: (1968, 2000, 2180), vsync: (1083, 1088, 1130), hskew: 0, vscan: 0, vrefresh: 144, mode_type: ModeTypeFlags(PREFERRED | DRIVER) }
2024-04-30T10:50:17.772411Z DEBUG niri::backend::tty: set max bpc to 8
2024-04-30T10:50:17.776297Z DEBUG niri::niri: putting output eDP-1 at x=0 y=0
2024-04-30T10:50:17.776404Z  INFO niri: listening on Wayland socket: wayland-1
2024-04-30T10:50:17.776411Z  INFO niri: IPC listening on: /run/user/1000/niri.wayland-1.188059.sock
```

**Comment by YaLTeR** [comment]
Indeed, they build with all features. Well anyhow, as far as I can see this shouldn't affect the spawning. I am actually not sure how it's possible for it to leave a lingering process if there's no error.

**Comment by YaLTeR** [comment]
Also from the configs that you posted there's nothing that should repeatedly spawn processes either, so that's also weird..

**Comment by YaLTeR** [comment]
Maybe you can somehow figure out where those children are coming from?

**Comment by aditya23043** [comment]
I'll check
Thank you for your help!

**Comment by aditya23043** [comment]
niri-bin is working perfectly fine for me with only 1% RAM usage on btm and htop
However, the child processes are still showing up
But it doesn't matter now since its work fine
Thanks!

**Comment by YaLTeR** [comment]
Well that's really weird. There shouldn't be any difference I think.

**Comment by YaLTeR** [comment]
Wait I think I know what this is.. If it's building with all features then it's enabling the Tracy instrumentation that collects all data at runtime until it runs out of memory

**Comment by lpnh** [comment]
I can confirm this. I faced the same issue and had `niri` installed from the extra repo. Replacing it with the `git` version from the AUR solved the problem for me.

I'm not entirely sure, but it seems that the package was moved from the AUR to the extra repo two days ago. Perhaps that's when the `--all-features` flag was introduced?

**Comment by YaLTeR** [comment]
Probably. I emailed the maintainer about it.

**Comment by YaLTeR** [comment]
The package should be fixed in 0.1.5-4
